### ISSUE TITLE
Gitlab CI issue

### ISSUE THREAD

**Issue by muuvmuuv** [moonrepo/proto#646]
### What version?

0.41.7

### Which command?

proto install

### What happened?

I am using the following script to install proto in our Gitlab pipelines. Unfortunately it fails pulling node. Is there a recommend way of using proto in other CI env?

```yml
.workspace:
  image: debian:latest
  before_script: &install_pnpm
    - apt-get update -y && apt-get install -y git unzip gzip xz-utils curl
    - mkdir -p $HOME/proto &&
      curl -fsSL https://github.com/moonrepo/proto/releases/download/v0.41.7/proto_cli-x86_64-unknown-linux-gnu.tar.xz |
      tar -xJ -C $HOME/proto --strip-components=1
    - $HOME/proto/proto install
    - $HOME/proto/proto run pnpm config set store-dir .cache/pnpm-store
    - $HOME/proto/proto run pnpm --silent install
```

### Trace logs?

```
node Installing Node.js ~22
pnpm Installing pnpm ~9
node Downloading pre-built archive
pnpm Downloading pre-built archive
pnpm Unpacking archive
pnpm pnpm 9.12.3 installed!
Error: net::http
  × Failed to make HTTP request for https://nodejs.org/download/release/
  │ v22.10.0/node-v22.10.0-linux-x64.tar.xz.
  ├─▶ error decoding response body
  ├─▶ error reading a body from connection
  ╰─▶ stream error received: unexpected internal error encountered
```

### Operating system?

Linux

### Architecture?

x64

**Comment by muuvmuuv** [comment]
Ok that seemed to be a network error that just occured a lot of times. Maybe a load balancing issue on node side since I tried a lot of times to fix the pipeline.

Now I get:

```
Successfully installed 2 tools!
 Tool  Configured  Resolved  Installed                        Config                            
 node  >=22        22.10.0   /root/.proto/tools/node/22.10.0  /builds/m8finder/app/package.json 
 pnpm  >=9         9.12.3    /root/.proto/tools/pnpm/9.12.3   /builds/m8finder/app/package.json 
$ $HOME/proto/proto run pnpm --version
proto-run 0.41.7
$ $HOME/proto/proto run pnpm -- --version
Error:   × No such file or directory (os error 2)
```

**Comment by milesj** [comment]
Nice find! This _should_ fallback to `sh` shell when one could not be found.

If possible, the command that errors with "No such file or directory", can you run it with `PROTO_LOG=trace`. It will help to pinpoint where the error is coming frmo.

**Comment by milesj** [comment]
Yeah this would hard error if no shell is found: https://github.com/moonrepo/proto/blob/master/crates/cli/src/commands/setup.rs#L65

**Comment by muuvmuuv** [comment]
Maybe using this helper would work out? https://github.com/moonrepo/starbase/blob/master/crates/shell/src/shell.rs#L59

**Comment by milesj** [comment]
It does use that helper.

**Comment by muuvmuuv** [comment]
Ran it with the debug env. First is using the install command which could not find a shell and second the initial error when trying to setup it manually. Weird is that I now get another error as initially described ^^...

```
[DEBUG 2024-11-13 19:58:20.620] proto  Running proto v0.42.1  bin="/root/.proto/bin/proto" args=["setup", "--yes"] pid=3227
[TRACE 19:58:20.620] starbase::app  Running startup phase 
[DEBUG 19:58:20.620] proto_core::proto  Creating proto environment, detecting store  store="/root/.proto" home="/root"
[TRACE 19:58:20.620] starbase::app  Running analyze phase 
[DEBUG 19:58:20.620] proto::systems  Loading configuration in upwards-global mode  working_dir="/builds/m8finder/app"
[DEBUG 19:58:20.620] proto_core::proto_config  Loading .prototools  file="/builds/m8finder/app/.prototools"
[TRACE 19:58:20.620] starbase_utils::fs  Reading file  file="/builds/m8finder/app/.prototools"
[TRACE 19:58:20.620] schematic::config::loader  Loading partial configuration  config="ProtoConfig"
[TRACE 19:58:20.620] schematic::config::loader  Creating layer from source  config="ProtoConfig" source="<code>"
[TRACE 19:58:20.620] schematic::config::loader  Merging partial layers into a final result  config="ProtoConfig"
[DEBUG 19:58:20.621] proto_core::proto_config  Merging loaded configs with global 
[DEBUG 19:58:20.621] proto_core::proto_config  Merged 1 configs 
[DEBUG 19:58:20.621] proto_core::proto_config  Merging loaded configs without global 
[DEBUG 19:58:20.621] proto_core::proto_config  Merged 1 configs 
[TRACE 19:58:20.621] starbase::app  Running execute phase 
[DEBUG 19:58:20.621] proto::commands::setup  Determining the shell to use 
[DEBUG 19:58:20.621] starbase_shell::shell  Attempting to detect the current shell 
[DEBUG 19:58:20.621] starbase_shell::shell  Detecting from operating system 
[DEBUG 19:58:20.622] starbase_shell::shell  Could not detect a shell! 
[TRACE 19:58:20.622] starbase::app  Running shutdown phase (because another phase failed) 
[TRACE 19:58:20.622] starbase::app  Error: Could not detect your terminal shell. Scanned the $SHELL environment variable and parent processes. 
[TRACE 19:58:20.622] starbase::app  Setting exit code  code=1
Error: shell::undetected
  × Could not detect your terminal shell. Scanned the $SHELL environment
  │ variable and parent processes.
```

```
$ PROTO_LOG=trace $HOME/proto/proto run pnpm install
[DEBUG 2024-11-13 20:11:38.135] proto  Running proto v0.42.1  bin="/root/proto/proto" args=["run", "pnpm", "install"] pid=3259
[TRACE 20:11:38.135] starbase::app  Running startup phase 
[DEBUG 20:11:38.135] proto_core::proto  Creating proto environment, detecting store  store="/root/.proto" home="/root"
[TRACE 20:11:38.135] starbase::app  Running analyze phase 
[DEBUG 20:11:38.135] proto::systems  Loading configuration in upwards-global mode  working_dir="/builds/m8finder/app"
[DEBUG 20:11:38.135] proto_core::proto_config  Loading .prototools  file="/builds/m8finder/app/.prototools"
[TRACE 20:11:38.135] starbase_utils::fs  Reading file  file="/builds/m8finder/app/.prototools"
[TRACE 20:11:38.135] schematic::config::loader  Loading partial configuration  config="ProtoConfig"
[TRACE 20:11:38.135] schematic::config::loader  Creating layer from source  config="ProtoConfig" source="<code>"
[TRACE 20:11:38.135] schematic::config::loader  Merging partial layers into a final result  config="ProtoConfig"
[DEBUG 20:11:38.136] proto_core::proto_config  Loading .prototools  file="/root/.proto/.prototools"
[TRACE 20:11:38.136] starbase_utils::fs  Opening file  file="/root/.proto/.prototools"
[TRACE 20:11:38.136] starbase_utils::fs_lock  Locking file  file="/root/.proto/.prototools"
[TRACE 20:11:38.136] starbase_utils::fs_lock  Unlocking file  file="/root/.proto/.prototools"
[TRACE 20:11:38.136] schematic::config::loader  Loading partial configuration  config="ProtoConfig"
[TRACE 20:11:38.136] schematic::config::loader  Creating layer from source  config="ProtoConfig" source="<code>"
[TRACE 20:11:38.136] schematic::config::loader  Merging partial layers into a final result  config="ProtoConfig"
[DEBUG 20:11:38.136] proto_core::proto_config  Merging loaded configs with global 
[DEBUG 20:11:38.136] proto_core::proto_config  Merged 2 configs 
[DEBUG 20:11:38.136] proto_core::proto_config  Merging loaded configs without global 
[DEBUG 20:11:38.136] proto_core::proto_config  Merged 1 configs 
[TRACE 20:11:38.136] starbase::app  Running execute phase 
[DEBUG 20:11:38.137] proto_core::tool_loader  Finding a configured plugin  tool="pnpm"
[DEBUG 20:11:38.137] proto_core::tool_loader  Using a built-in plugin  plugin="https://github.com/moonrepo/tools/releases/download/node_depman_tool-v0.14.0/node_depman_tool.wasm"
[TRACE 20:11:38.137] warpgate::loader  Creating plugin loader  cache_dir="/root/.proto/plugins"
[TRACE 20:11:38.137] warpgate::loader  Loading plugin pnpm  id="pnpm" locator="https://github.com/moonrepo/tools/releases/download/node_depman_tool-v0.14.0/node_depman_tool.wasm"
[TRACE 20:11:38.137] starbase_utils::fs  Reading file metadata  file="/root/.proto/plugins/pnpm-90d98126ced2031cb2509881396dd3d1481f6b239d032ba37787b9dc616a011d.wasm"
[TRACE 20:11:38.137] warpgate::loader  Plugin already downloaded and cached  id="pnpm" path="/root/.proto/plugins/pnpm-90d98126ced2031cb2509881396dd3d1481f6b239d032ba37787b9dc616a011d.wasm"
[DEBUG 20:11:38.137] proto_core::tool_loader  Loading WASM plugin  source="/root/.proto/plugins/pnpm-90d98126ced2031cb2509881396dd3d1481f6b239d032ba37787b9dc616a011d.wasm"
[TRACE 20:11:38.140] warpgate::plugin  Storing plugin identifier  id="pnpm"
[TRACE 20:11:38.140] warpgate::plugin  Storing host environment  id="pnpm" env={"arch":"x64","libc":"gnu","os":"linux","home_dir":{"path":"/userhome/","virtual_prefix":"/userhome","real_prefix":"/root"}}
[DEBUG 20:11:38.140] proto_core::tool  Creating tool pnpm and instantiating plugin 
[DEBUG 20:11:38.140] warpgate::client  Creating HTTP client 
[DEBUG 20:11:38.149] warpgate::client  Created HTTP client 
[TRACE 20:11:38.149] warpgate::plugin  Creating plugin container  id="pnpm"
[TRACE 20:11:38.171] warpgate::plugin  Created plugin container  id="pnpm" plugin="267977d0-54b2-4874-bd03-d8d55b26e9fc"
[DEBUG 20:11:38.171] proto_core::tool  Created tool pnpm and its WASM runtime 
[TRACE 20:11:38.171] warpgate::plugin  Calling guest function register_tool  id="pnpm" plugin="267977d0-54b2-4874-bd03-d8d55b26e9fc" input={"id":"pnpm"}
[TRACE 20:11:38.172] warpgate::plugin  Called guest function register_tool  id="pnpm" plugin="267977d0-54b2-4874-bd03-d8d55b26e9fc" output={"config_schema":{"name":"NodeDepmanPluginConfig","ty":{"type":"Struct","fields":{"dist-url":{"schema":{"ty":{"type":"String"}},"optional":true},"shared-globals-dir":{"schema":{"ty":{"type":"Boolean"}},"optional":true}},"partial":false}},"inventory":{},"minimum_proto_version":"0.42.0","name":"pnpm","plugin_version":"0.14.0","type":"DependencyManager"}
[DEBUG 20:11:38.172] proto_core::tool_manifest  Loading manifest.json  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 20:11:38.172] starbase_utils::fs  Opening file  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 20:11:38.172] starbase_utils::fs_lock  Locking file  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 20:11:38.172] starbase_utils::fs_lock  Unlocking file  file="/root/.proto/tools/pnpm/manifest.json"
[DEBUG 20:11:38.172] proto_core::version_detector  Using explicit version passed on the command line  tool="pnpm" version=Alias("install")
[DEBUG 20:11:38.172] proto_core::flow::resolve  Resolving a semantic version or alias  tool="pnpm" initial_version="install"
[DEBUG 20:11:38.172] proto_core::flow::resolve  Loading available versions  tool="pnpm"
[TRACE 20:11:38.173] starbase_utils::fs  Reading file metadata  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 20:11:38.173] starbase_utils::fs  Reading file  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 20:11:38.173] starbase_utils::json  Reading JSON file  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 20:11:38.173] warpgate::plugin  Calling guest function resolve_version  id="pnpm" plugin="267977d0-54b2-4874-bd03-d8d55b26e9fc" input={"initial":"install"}
[TRACE 20:11:38.173] warpgate::plugin  Called guest function resolve_version  id="pnpm" plugin="267977d0-54b2-4874-bd03-d8d55b26e9fc" output={}
[TRACE 20:11:38.173] proto_core::version_resolver  Resolving a version with manifest  candidate="install"
[TRACE 20:11:38.173] proto_core::version_resolver  Found an alias, resolving further  alias="install"
[TRACE 20:11:38.173] proto_core::version_resolver  Alias does not exist, trying others  alias="install"
[TRACE 20:11:38.174] starbase::app  Running shutdown phase (because another phase failed) 
[TRACE 20:11:38.174] starbase::app  Error: Failed to resolve install to a valid supported version for pnpm. 
[TRACE 20:11:38.174] starbase::app  Setting exit code  code=1
Error: proto::version::unresolved
  × Failed to resolve install to a valid supported version for pnpm.
  help: Does this version exist and has it been released?
```

**Comment by milesj** [comment]
@muuvmuuv It's wild it couldn't find any shell. I guess I'll update it to fallback to something instead of hard erroring.

As for the second error, it needs to be `proto run pnpm -- install`. Args passed down need to be separated by `--`.

**Comment by muuvmuuv** [comment]
Oh yes, so here is the original error with verbose logging:

```
$ PROTO_LOG=trace $HOME/proto/proto run pnpm -- install
[DEBUG 2024-11-14 07:12:54.127] proto  Running proto v0.41.7  bin="/root/proto/proto" args=["run", "pnpm", "--", "install"] pid=3259
[TRACE 07:12:54.127] starbase::app  Running startup phase 
[DEBUG 07:12:54.127] proto_core::proto  Creating proto environment, detecting store  store="/root/.proto" home="/root"
[TRACE 07:12:54.127] starbase::app  Running analyze phase 
[DEBUG 07:12:54.127] proto::systems  Loading configuration in upwards-global mode  working_dir="/builds/m8finder/app"
[DEBUG 07:12:54.127] proto_core::proto_config  Loading .prototools  file="/builds/m8finder/app/.prototools"
[TRACE 07:12:54.127] starbase_utils::fs  Reading file  file="/builds/m8finder/app/.prototools"
[TRACE 07:12:54.127] schematic::config::loader  Loading partial configuration  config="ProtoConfig"
[TRACE 07:12:54.127] schematic::config::loader  Creating layer from source  config="ProtoConfig" source="<code>"
[TRACE 07:12:54.128] schematic::config::loader  Merging partial layers into a final result  config="ProtoConfig"
[DEBUG 07:12:54.128] proto_core::proto_config  Loading .prototools  file="/root/.proto/.prototools"
[TRACE 07:12:54.128] starbase_utils::fs  Opening file  file="/root/.proto/.prototools"
[TRACE 07:12:54.128] starbase_utils::fs_lock  Locking file  file="/root/.proto/.prototools"
[TRACE 07:12:54.128] starbase_utils::fs_lock  Unlocking file  file="/root/.proto/.prototools"
[TRACE 07:12:54.128] schematic::config::loader  Loading partial configuration  config="ProtoConfig"
[TRACE 07:12:54.128] schematic::config::loader  Creating layer from source  config="ProtoConfig" source="<code>"
[TRACE 07:12:54.128] schematic::config::loader  Merging partial layers into a final result  config="ProtoConfig"
[DEBUG 07:12:54.128] proto_core::proto_config  Merging loaded configs with global 
[DEBUG 07:12:54.128] proto_core::proto_config  Merged 2 configs 
[DEBUG 07:12:54.129] proto_core::proto_config  Merging loaded configs without global 
[DEBUG 07:12:54.129] proto_core::proto_config  Merged 1 configs 
[TRACE 07:12:54.129] starbase::app  Running execute phase 
[DEBUG 07:12:54.129] proto_core::tool_loader  Finding a configured plugin  tool="pnpm"
[DEBUG 07:12:54.129] proto_core::tool_loader  Using a built-in plugin  plugin="https://github.com/moonrepo/tools/releases/download/node_depman_tool-v0.13.1/node_depman_tool.wasm"
[TRACE 07:12:54.129] warpgate::loader  Creating plugin loader  cache_dir="/root/.proto/plugins"
[TRACE 07:12:54.129] warpgate::loader  Loading plugin pnpm  id="pnpm" locator="https://github.com/moonrepo/tools/releases/download/node_depman_tool-v0.13.1/node_depman_tool.wasm"
[TRACE 07:12:54.130] starbase_utils::fs  Reading file metadata  file="/root/.proto/plugins/pnpm-b9341305d212d01f814c24338dadc08498000fcc4a6b6e8e1e7b27fc5cd7478c.wasm"
[TRACE 07:12:54.130] warpgate::loader  Plugin already downloaded and cached  id="pnpm" path="/root/.proto/plugins/pnpm-b9341305d212d01f814c24338dadc08498000fcc4a6b6e8e1e7b27fc5cd7478c.wasm"
[DEBUG 07:12:54.130] proto_core::tool_loader  Loading WASM plugin  source="/root/.proto/plugins/pnpm-b9341305d212d01f814c24338dadc08498000fcc4a6b6e8e1e7b27fc5cd7478c.wasm"
[TRACE 07:12:54.132] warpgate::plugin  Storing plugin identifier  id="pnpm"
[TRACE 07:12:54.132] warpgate::plugin  Storing host environment  id="pnpm" env={"arch":"x64","libc":"gnu","os":"linux","home_dir":{"path":"/userhome/","virtual_prefix":"/userhome","real_prefix":"/root"}}
[DEBUG 07:12:54.132] proto_core::tool  Creating tool pnpm and instantiating plugin 
[DEBUG 07:12:54.132] warpgate::client  Creating HTTP client 
[DEBUG 07:12:54.141] warpgate::client  Created HTTP client 
[TRACE 07:12:54.141] warpgate::plugin  Creating plugin container  id="pnpm"
[TRACE 07:12:54.164] warpgate::plugin  Created plugin container  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba"
[DEBUG 07:12:54.164] proto_core::tool  Created tool pnpm and its WASM runtime 
[TRACE 07:12:54.164] warpgate::plugin  Calling guest function register_tool  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" input={"id":"pnpm"}
[TRACE 07:12:54.164] warpgate::plugin  Called guest function register_tool  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" output={"config_schema":{"name":"NodeDepmanPluginConfig","ty":{"type":"Struct","fields":{"dist-url":{"schema":{"ty":{"type":"String"}},"hidden":false,"nullable":false,"optional":true,"read_only":false,"write_only":false},"shared-globals-dir":{"schema":{"ty":{"type":"Boolean"}},"hidden":false,"nullable":false,"optional":true,"read_only":false,"write_only":false}},"partial":false}},"inventory":{},"name":"pnpm","plugin_version":"0.13.1","type":"DependencyManager"}
[DEBUG 07:12:54.164] proto_core::tool_manifest  Loading manifest.json  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 07:12:54.164] starbase_utils::fs  Opening file  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 07:12:54.164] starbase_utils::fs_lock  Locking file  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 07:12:54.164] starbase_utils::fs_lock  Unlocking file  file="/root/.proto/tools/pnpm/manifest.json"
[TRACE 07:12:54.165] proto_core::version_detector  Attempting to find version from .prototools files  tool="pnpm"
[DEBUG 07:12:54.165] proto_core::version_detector  Detected version from .prototools file  tool="pnpm" version="~9" file="/builds/m8finder/app/.prototools"
[DEBUG 07:12:54.165] proto_core::flow::resolve  Resolving a semantic version or alias  tool="pnpm" initial_version="~9"
[DEBUG 07:12:54.165] proto_core::flow::resolve  Loading available versions  tool="pnpm"
[TRACE 07:12:54.165] starbase_utils::fs  Reading file metadata  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 07:12:54.165] starbase_utils::fs  Reading file  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 07:12:54.165] starbase_utils::json  Reading JSON file  file="/root/.proto/tools/pnpm/remote-versions.json"
[TRACE 07:12:54.166] warpgate::plugin  Calling guest function resolve_version  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" input={"initial":"~9"}
[TRACE 07:12:54.166] warpgate::plugin  Called guest function resolve_version  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" output={}
[TRACE 07:12:54.166] proto_core::version_resolver  Resolving a version with manifest  candidate="~9"
[TRACE 07:12:54.166] proto_core::version_resolver  Found a requirement, resolving further  requirement="~9"
[TRACE 07:12:54.166] proto_core::version_resolver  Resolved to locally installed version  version="9.13.0"
[DEBUG 07:12:54.166] proto_core::flow::resolve  Resolved to 9.13.0  tool="pnpm" version="9.13.0"
[DEBUG 07:12:54.166] proto_core::flow::setup  Checking if tool is installed  tool="pnpm" install_dir="/root/.proto/tools/pnpm/9.13.0"
[DEBUG 07:12:54.166] proto_core::flow::setup  Tool has already been installed, locating binaries and shims  tool="pnpm" install_dir="/root/.proto/tools/pnpm/9.13.0"
[TRACE 07:12:54.166] warpgate::plugin  Calling guest function locate_executables  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" input={"context":{"proto_version":"0.41.7","tool_dir":{"path":"/proto/tools/pnpm/9.13.0","virtual_prefix":"/proto","real_prefix":"/root/.proto"},"version":"9.13.0"}}
[TRACE 07:12:54.167] warpgate::plugin  Called guest function locate_executables  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" output={"exes_dir":".","globals_lookup_dirs":["$PREFIX/bin","$PNPM_HOME","$HOME/.local/share/pnpm"],"primary":{"exe_path":"bin/pnpm.cjs","no_bin":true,"parent_exe_name":"node"},"secondary":{"pnpx":{"no_bin":true,"shim_before_args":"dlx"}}}
[DEBUG 07:12:54.167] proto_core::flow::locate  Locating primary executable for tool  tool="pnpm"
[DEBUG 07:12:54.168] proto_core::flow::locate  Found an executable  tool="pnpm" exe_path="/root/.proto/tools/pnpm/9.13.0/bin/pnpm.cjs"
[DEBUG 07:12:54.168] proto_core::flow::locate  Locating globals directories for tool  tool="pnpm"
[DEBUG 07:12:54.168] proto_core::flow::locate  Located possible globals directories  tool="pnpm" dirs=["/root/.local/share/pnpm"]
[TRACE 07:12:54.168] warpgate::plugin  Calling guest function pre_run  id="pnpm" plugin="b414d1ad-1fa8-48ad-9f17-919711f20cba" input={"context":{"proto_version":"0.41.7","tool_dir":{"path":"/proto/tools/pnpm/9.13.0","virtual_prefix":"/proto","real_prefix":"/root/.proto"},"version":"9.13.0"},"globals_dir":null,"globals_prefix":null,"passthrough_args":["install"]}
[TRACE 07:12:54.168] warpgate::plugin  Called guest function pre_run  id="pnpm" plugin="b[414](https://gitlab.corp.m8finder.net/m8finder/app/-/jobs/7649#L414)d1ad-1fa8-48ad-9f17-919711f20cba" output={"args":null,"env":null}
[DEBUG 07:12:54.168] proto::commands::run  Running pnpm  bin="node" args=["/root/.proto/tools/pnpm/9.13.0/bin/pnpm.cjs", "install"] pid=3259
[TRACE 07:12:54.168] starbase_utils::fs  Writing file  file="/root/.proto/tools/pnpm/9.13.0/.last-used"
[TRACE 07:12:54.170] starbase::app  Running shutdown phase (because another phase failed) 
[TRACE 07:12:54.170] starbase::app  Setting exit code  code=1
Error:   × No such file or directory (os error 2)
```

**Comment by muuvmuuv** [comment]
I just tested within bare debian image (which I use in GitLab pipeline) and $SHELL is defined there. I wonder if GitLab does something weird under the hood. I will research on that.

**Comment by milesj** [comment]
@muuvmuuv 

Hrmm yeah something doesn't seem right. Let me dig into this a bit more.

As for that follow-up pnpm error, I had a crazy thought. Does `node` exist? I wonder if the "No such file or directory" error happens because it's trying to execute a node binary that does not exist.

**Comment by muuvmuuv** [comment]
Maybe something for another ticket, but how do I cache proto properly in CI, is it enough to put PROTO_HOME in the cache? I tried that but it still shows me "Downloading pre-build binaries".

<details>
<summary>Here is my .gitlab-ci.yml (you can test that with gitlab-ci-local tool):</summary>

```yml
variables:
  CACHE_KEY: $CI_COMMIT_REF_SLUG
  PROTO_HOME: .proto
  GIT_DEPTH: 69

stages:
  - lint
  - test
  - build

.cache: &cache
  - key: &cache-key
      files:
        - .prototools
        - pnpm-lock.yaml
    policy: pull
    paths: &cache-paths
      - $PROTO_HOME
      - .cache

.workspace:
  image: debian:latest
  before_script:
    - apt-get -qq update -y && apt-get install -y -qq -o=Dpkg::Use-Pty=0 git unzip gzip xz-utils curl
    - curl -fsSL https://moonrepo.dev/install/proto.sh | PROTO_HOME=$PROTO_HOME bash -s -- latest --shell=bash --yes
    - export PATH="$PROTO_HOME/shims:$PROTO_HOME/bin:$PATH"
    - proto install
    - pnpm config set store-dir .cache/pnpm-store
    - pnpm --silent install
  cache: *cache

# Make sure everything is cached

modules:
  stage: .pre
  extends: .workspace
  script: echo 'Done'
  cache:
    - key: *cache-key
      policy: pull-push
      paths: *cache-paths
```

</details>

**Comment by milesj** [comment]
@muuvmuuv For caching, you should only cache the `~/.proto/tools` directory. The other folders/files aren't important as they'll be created on demand.

You would need a way to bust the cache though. In GitHub we use the `.prototools` file and some other params: https://github.com/moonrepo/setup-toolchain/blob/master/helpers.ts#L100

**Comment by muuvmuuv** [comment]
Yeah. But GitLab does not allow to cache outside build-dir, so it must be something like `./.proto`. I tried setting `PROTO_HOME` but the proto setup script seems to still write ~/.proto into bashrc. Tried it that way:
```
modules      $ export PROTO_HOME=.cache/proto
modules      $ curl -fsSL https://moonrepo.dev/install/proto.sh | PROTO_HOME=.cache/proto PROTO_DEBUG=true bash -s -- latest --shell=bash --yes
######################################################################## 100.0%              
modules      > Finishing proto installation...
modules      > Successfully installed proto to /gcl-builds/.cache/proto/bin/proto!
modules      > The shell profile at ~/.bashrc was updated.
modules      > Launch a new terminal window to start using proto!
modules      > 
modules      > Need help? Join our Discord https://discord.gg/qCh9MEynv2
modules      > 
modules      > arch=Linux aarch64
modules      > target=proto_cli-aarch64-unknown-linux-gnu
modules      > download_url=https://github.com/moonrepo/proto/releases/latest/download/proto_cli-aarch64-unknown-linux-gnu.tar.xz
modules      > bin_path=.cache/proto/bin/proto
modules      > shim_path=.cache/proto/bin/proto-shim
modules      > is_wsl=false
modules      > deps=ldd (Debian GLIBC 2.36-9+deb12u8) 2.36
modules      > Copyright (C) 2022 Free Software Foundation, Inc.
modules      > This is free software; see the source for copying conditions.  There is NO
modules      > warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
modules      > Written by Roland McGrath and Ulrich Drepper.
modules      $ source ~/.bashrc
modules      $ cat ~/.bashrc
modules      > # ~/.bashrc: executed by bash(1) for non-login shells.
modules      > 
modules      > # Note: PS1 and umask are already set in /etc/profile. You should not
modules      > # need this unless you want different defaults for root.
modules      > # PS1='${debian_chroot:+($debian_chroot)}\h:\w\$ '
modules      > # umask 022
modules      > 
modules      > # You may uncomment the following lines if you want `ls' to be colorized:
modules      > # export LS_OPTIONS='--color=auto'
modules      > # eval "$(dircolors)"
modules      > # alias ls='ls $LS_OPTIONS'
modules      > # alias ll='ls $LS_OPTIONS -l'
modules      > # alias l='ls $LS_OPTIONS -lA'
modules      > #
modules      > # Some more alias to avoid making mistakes:
modules      > # alias rm='rm -i'
modules      > # alias cp='cp -i'
modules      > # alias mv='mv -i'
modules      > 
modules      > # proto
modules      > export PROTO_HOME="$HOME/.proto";
modules      > export PATH="$PROTO_HOME/shims:$PROTO_HOME/bin:$PATH";<gclShellPromptPlaceholder> proto install
modules      > /gcl-cmd: line 15: proto: command not found
modules      finished in 7.28 s  FAIL 127 
```

**Comment by milesj** [comment]
Ok yeah that's just a big. The setup script is hard-coded, let me fix that.

**Comment by milesj** [comment]
Fixed the shell and setup issues in v0.43
